### PR TITLE
ffmpeg{,_{4,5}}-headless: init

### PR DESCRIPTION
**Pull request by vs49688**
###### Description of changes

The FFmpeg closure is huge. It pulls in SDL, vdpau, and pulseaudio, which in turn pull in pipewire, dbus, etc., most of which are only useful for desktop systems.

This introduces `-minimal` variants of FFmpeg that are quite a bit smaller (not small enough imo), and updates Navidrome and unpaper to use them. There are probably more packages that could benefit from this.

For comparison:
```
$ nix path-info -Sh .#ffmpeg_4 .#ffmpeg_4-minimal .#ffmpeg_5 .#ffmpeg_5-minimal
/nix/store/0p929r2ifvymzd7jjny4gfmfa68k4g82-ffmpeg-4.4.2-bin     562.2M
/nix/store/a95pgdp816i9h230pa3yh750ar9vbdd3-ffmpeg-4.4.2-man       1.2M
/nix/store/ba3h2sa7m3kfa55w50dbdkb50wqia25f-ffmpeg-4.4.2-bin     163.6M
/nix/store/bspcw6q45bzyizvmkzd8g0zg5brhp0zj-ffmpeg-5.1.2-bin     562.6M
/nix/store/fax05c69zn6jg34n0qi5syghby0v4cwj-ffmpeg-5.1.2-man       1.3M
/nix/store/mxsqs9vk95q0cakrpj3c481wzkqacfmi-ffmpeg-4.4.2-man       1.2M
/nix/store/sjdp7cbnp6yf0hzflfalzcc47bph1xwn-ffmpeg-5.1.2-bin     164.0M
/nix/store/w0ra8kmiyn0l8xvfi5g1lb9a8vpdhppa-ffmpeg-5.1.2-man       1.3M
```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vs49688** [comment]
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>paperless</li>
    <li>paperless-ng</li>
  </ul>
</details>
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>ffmpeg-minimal</li>
    <li>ffmpeg_5-minimal</li>
    <li>gscan2pdf</li>
    <li>navidrome</li>
    <li>ocrmypdf (python310Packages.ocrmypdf)</li>
    <li>paperless-ngx</li>
    <li>pdfsandwich</li>
    <li>python39Packages.ocrmypdf</li>
    <li>unpaper</li>
  </ul>
</details>

Slight oddity - `paperless` and `paperless-ng` fail, but they're just aliases to `paperless-ngx`, which succeeded?

**Comment by risicle** [comment]
Nice idea. I wonder if what's really needed is a "headless" version with no ui support at all.

**Comment by vs49688** [comment]
> Nice idea. I wonder if what's really needed is a "headless" version with no ui support at all.

That's basically what I'm going for. I'm still experimenting with it - I'm compiling with `libva-minimal` currently.

**Comment by vs49688** [comment]
After switching to `libva-minimal`, there's no longer any x11/wayland dependencies. This is about as small as its going to get unless I start removing headless features.

```
/nix/store/9dyffcdc3wd7b8v0jgjpsvcwavrvzr04-ffmpeg-5.1.2-man	   1.3M
/nix/store/fxfwfbi4li1z4x9ppc4igi4vpbdayjgp-ffmpeg-4.4.2-bin	 156.5M
/nix/store/h13zc0ds6af1fix1zs45w0m0k7rrblgn-ffmpeg-4.4.2-man	   1.2M
/nix/store/ry2rg145vr982vd8fr5qm2cg6xvqk0qk-ffmpeg-5.1.2-bin	 156.9M
```

**Comment by Atemu** [comment]
What do you think of calling it `headless` instead? 

Essentially what you're building is a ffmpeg which doesn't come with dependencies that are only useful on a headed system rather than one that doesn't come with any, are you?

This would also be helpful for SDL bootstrap.

**Comment by vs49688** [comment]
> What do you think of calling it `headless` instead?
> 

No problem at all. Done.

> Essentially what you're building is a ffmpeg which doesn't come with dependencies that are only useful on a headed system rather than one that doesn't come with any, are you?

Yep. My main motivation was to reduce the closure size to help my poor 30gb "server".

**Comment by vs49688** [comment]
> This would also be helpful for SDL bootstrap.

Pipewire too, but that should probably happen after this is merged.
```
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index d1f8b6bf315..f7378003d21 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -15649,12 +15649,8 @@ with pkgs;

   pipewire = callPackage ../development/libraries/pipewire {
     # ffmpeg depends on SDL2 which depends on pipewire by default.
-    # Break the cycle by disabling pipewire support in our ffmpeg.
-    ffmpeg = ffmpeg.override {
-      SDL2 = SDL2.override {
-        pipewireSupport = false;
-      };
-    };
+    # Break the cycle by depending on ffmpeg-headless.
+    ffmpeg = ffmpeg-headless;
   };
```

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -18421,9 +18421,28 @@ with pkgs;
...
     inherit (darwin.apple_sdk.frameworks) Cocoa CoreMedia VideoToolbox;
   };
+
+  ffmpeg_5-headless = callPackage ../development/libraries/ffmpeg/5.nix {
+    inherit (darwin.apple_sdk.frameworks) Cocoa CoreMedia VideoToolbox;
+
+    sdlSupport = false;
+    vdpauSupport = false;
+    pulseaudioSupport = false;
+    libva = libva-minimal;
```
Comment (Atemu) [inline pkgs/top-level/all-packages.nix:18443]: Have you tested whether libva still works?

**Comment by Atemu** [comment]
Yeah, SDL, pipewire and what have you will need to go to staging in a separate PR.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (vs49688) [reply pkgs/top-level/all-packages.nix:18443]: Yep:
```
[AVHWDeviceContext @ 0x18484c0] Trying to use DRM render node for device 0.
[AVHWDeviceContext @ 0x18484c0] libva: VA-API version 1.15.0
[AVHWDeviceContext @ 0x18484c0] libva: Trying to open /run/opengl-driver/lib/dri/iHD_drv_video.so
[AVHWDeviceContext @ 0x18484c0] libva: Found init function __vaDriverInit_1_15
[AVHWDeviceContext @ 0x18484c0] libva: va_openDriver() returns 0
[AVHWDeviceContext @ 0x18484c0] Initialised VAAPI connection: version 1.15
[AVHWDeviceContext @ 0x18484c0] Format 0x41524742 -> bgra.
[AVHWDeviceContext @ 0x18484c0] Format 0x42475241 -> argb.
[AVHWDeviceContext @ 0x18484c0] Format 0x41424752 -> rgba.
...
...
[h264 @ 0x190d640] nal_unit_type: 8(PPS), nal_ref_idc: 3
[h264 @ 0x190d640] nal_unit_type: 5(IDR), nal_ref_idc: 3
[h264 @ 0x190d640] Format vaapi chosen by get_format().
[h264 @ 0x190d640] Format vaapi requires hwaccel initialisation.
[h264 @ 0x190d640] Considering format 0x3231564e -> nv12.
[h264 @ 0x190d640] Picked nv12 (0x3231564e) as best match for yuv420p.
[AVHWFramesContext @ 0x7f26b8049900] Created surface 0.
[AVHWFramesContext @ 0x7f26b8049900] Direct mapping possible.
[AVHWFramesContext @ 0x7f26b8049900] Created surface 0x1.
[AVHWFramesContext @ 0x7f26b8049900] Created surface 0x2.
[AVHWFramesContext @ 0x7f26b8049900] Created surface 0x3.
```

`./result-bin/bin/ffmpeg -loglevel trace -hwaccel vaapi -i ~/Videos/h264test.mp4 -f null -`

Comment (Atemu) [reply pkgs/top-level/all-packages.nix:18443]: Awesome, thank you!

**Review by Atemu — APPROVED** [review]

**Comment by risicle** [comment]
`nixpkgs-review` happy, macos 10.15

**Comment by risicle** [comment]
`pkgsMusl` and `pkgsCross.aarch64-multiplatform` variants build too, for both 4 and 5.

**Pull request merged**